bio/bio.go

```diff
@@ -0,0 +1,323 @@
+// all the input from the underlying io.Reader before exiting.
+func (p *Parser[DataType, DataTypeHeader]) ParseWithHeader() ([]*DataType, *DataTypeHeader, error) {
+	header, headerErr := p.Header()
+	data, err := p.Parse()
+	if headerErr != nil {
+		return data, header, err
+	}
+	if err != nil {
+		return data, header, err
+	}
+	return data, header, nil
+}
+
+/******************************************************************************
+
+Concurrent higher-level functions
+
+******************************************************************************/
+
+// ParseToChannel pipes all records/reads/lines from a parser into a channel,
+// then optionally closes that channel. If parsing a single file,
+// "keepChannelOpen" should be set to false, which will close the channel once
+// parsing is complete. If many files are being parsed to a single channel,
+// keepChannelOpen should be set to true, so that an external function will
+// close channel once all are done parsing.
+//
+// Context can be used to close the parser in the middle of parsing - for
+// example, if an error is found in another parser elsewhere and all files
+// need to close.
+func (p *Parser[DataType, DataTypeHeader]) ParseToChannel(ctx context.Context, channel chan<- DataType, keepChannelOpen bool) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			record, err := p.Next()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					err = nil // EOF not treated as parsing error.
+				}
+				if record != nil {
+					channel <- *record
+				}
+				if !keepChannelOpen {
+					close(channel)
+				}
+				return err
+			}
+			channel <- *record
+		}
+	}
+}
+
+// ManyFastaToChannel takes a list of FASTA parsers and pipes all
+// records/reads/lines to a single channel.
+func ManyFastaToChannel(ctx context.Context, parsers []*Parser[fasta.Record, fasta.Header], channel chan<- fasta.Record) error {
+	return manyToChannel[fasta.Record, fasta.Header](ctx, parsers, channel)
+}
+
+// ManyFastqToChannel takes a list of FASTQ parsers and pipes all
+// records/reads/lines to a single channel.
+func ManyFastqToChannel(ctx context.Context, parsers []*Parser[fastq.Read, fastq.Header], channel chan<- fastq.Read) error {
+	return manyToChannel[fastq.Read, fastq.Header](ctx, parsers, channel)
+}
+
+// ManySlow5ToChannel takes a list of SLOW5 parsers and pipes all
+// records/reads/lines to a single channel.
+func ManySlow5ToChannel(ctx context.Context, parsers []*Parser[slow5.Read, slow5.Header], channel chan<- slow5.Read) error {
+	return manyToChannel[slow5.Read, slow5.Header](ctx, parsers, channel)
+}
+
+// ManyPileupToChannel takes a list of pileup parsers and pipes all
+// records/reads/lines to a single channel.
+func ManyPileupToChannel(ctx context.Context, parsers []*Parser[pileup.Line, pileup.Header], channel chan<- pileup.Line) error {
+	return manyToChannel[pileup.Line, pileup.Header](ctx, parsers, channel)
+}
```

Comment:
`ParseToChannel` already checks for `ctx.Done()` so this is unnecessary.

Reply:
True. Deleted.